### PR TITLE
Add in config blocks with validation

### PR DESCRIPTION
**Pull request by jsallay**
Signed-off-by: John Sallay <jasallay@gmail.com>

# Pull Request Details

## Description

It is frequently useful to use a config file to set the parameters for a flowgraph.  In GRC, a variable config block can be used, but there are several shortcomings.  Each config parameters in read in separately and requires a different block to be instantiated.  It also requires that a default value is specified, even if there isn't a sensible default.

I added in a few blocks, that, I believe, add to a cleaner user experience and allow for validation of the config file.  A json schema file can be used to ensure that all required fields are present and have valid values.  The config file is available in GRC as a dictionary, which can be used directly in parameters for other blocks.  i.e specify the Sample Rate as config["samp_rate"].

## Testing Done

I built my branch from source and  manually tested in GRC.  I created sample json, yaml, and schema files.  I created a simple flowgraph and ensured that it worked as expected in GRC and that the python produced was valid and worked.  Note that these blocks could be extended to work in c++, but that would be a lot more effort and introduce new dependencies.  These blocks are presently python only.

## Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [ ] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [ ] I have added tests to cover my changes, and all previous tests pass.

### CODE REVIEW

**Comment by 777arc** [comment]
Cool idea!

**Comment by mormj** [comment]
Love this.

File: grc/blocks/json_config.block.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+id: json_config
+label: Json Config
```
Comment (777arc) [inline grc/blocks/json_config.block.yml:2]: JSON Config

**Comment by 777arc** [comment]
Because all built-in blocks have a link to a wiki page, see below, that's the best spot to stick any usage related docs and example flowgraphs, although the yaml file should still explain what the block does and how any params are used.  I would update the Json > JSON before we create the wiki page though, because the wiki page titles match the label.

Screenshot from 2022-02-26 14-49-03

File: grc/blocks/json_config.block.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+flags: [ show_id, python ]
+
+parameters:
+-   id: config_file
+    label: Config File
+    dtype: file_open
+    default: default
```
Comment (777arc) [inline grc/blocks/json_config.block.yml:9]: probably better off with an empty string here instead

File: grc/blocks/json_config.block.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        from jsonschema import validate
+    var_make: "with open(${config_file}) as fid:\n\
+              \    self.${id} = ${id} = json.load(fid)\n\
+              self.${id}_schema = ${id}_schema = ${schema_file}\n\
+              if ${id}_schema:\n\
+              \    with open(${id}_schema) as fid:\n\
+              \        validate(${id}, json.load(fid))"
+
+documentation: |-
+    This block represents a json config file, that is read in as a dictionary.
```
Comment (777arc) [inline grc/blocks/json_config.block.yml:34]: First time using the block, some questions pop in my head like 1) is the schema optional 2) is the json supposed to be only 1 level deep 3) you say it's read in as a dictionary but what does that mean in GR sense, like does each field become a variable? 

I think we should present people with the most basic example of using it, so for example:
1 - Make a new file called example.json with the contents:
{
  "amplitude": 0.1
}
2 - Add the JSON Config block and point Config File to that file
3 - Make a signal source, and for Amplitude enter the value json_config_0[ 'amplitude']
4 - Connect it to a Time Sink through a Throttle and you'll see the amplitude show up as 0.1

**Comment by 777arc** [comment]
Went ahead and started the wiki page https://wiki.gnuradio.org/index.php?title=JSON_Config it assumes Json gets swapped to JSON for the label

File: grc/blocks/json_config.block.yml (hunk 2 above)
Comment (jsallay) [reply grc/blocks/json_config.block.yml]: Done

File: grc/blocks/json_config.block.yml (hunk 3 above)
Comment (jsallay) [reply grc/blocks/json_config.block.yml]: Where should I put this example?  I could put it right in the doc string.  For example,
```
With an input file of 
{
  "samp_rate": 1e6,
  "center_freq": 8e9 
}
```
We can access the data, like `config["samp_rate"]`

Comment (777arc) [reply grc/blocks/json_config.block.yml]: see https://wiki.gnuradio.org/index.php?title=JSON_Config#Example_Flowgraph

Comment (777arc) [reply grc/blocks/json_config.block.yml]: Wouldn't hurt to also have an example json in the yaml as well, and it's fine if it's different example than the wiki

File: grc/blocks/yaml_config.block.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+id: yaml_config
+label: Yaml Config
```
Comment (777arc) [inline grc/blocks/yaml_config.block.yml:2]: YAML =)

**Comment by 777arc** [comment]
I love the YAML block, in fact I'll personally be more likely to use that than the JSON.  The wiki docs can always follow the merge so we can focus on the code portion for now.  I like the idea of each .yml file in the documentation section explaining what the 2 params do, and have an example json/yaml file and then say how to actually use the values in another block's param, that way the user doesn't have to guess at how it works.

File: grc/blocks/json_config.block.yml (hunk 3 above)
Comment (jsallay) [reply grc/blocks/json_config.block.yml]: Added in examples

**Review by mormj — APPROVED** [review]
Love it - merging!

**Comment by willcode** [comment]
Noticed that this imports from `jsonschema`. On my system, `python3-jsonschema` is only there because `docker-compose` requires it. Have we identified this as a requirement?

**Comment by mormj** [comment]
Good point - probably need to add a check to the cmake and make installation of the grc dependent on the output.  Creating an issue

**Pull request merged**